k8s_references.go

```diff
@@ -76,19 +76,34 @@ func KRef(namespace, name string) ObjectRef {
 	}
 }
 
-// KObjs returns slice of ObjectRef from an slice of ObjectMeta
-func KObjs(arg interface{}) []ObjectRef {
-	s := reflect.ValueOf(arg)
+type kobjs struct {
+	arg interface{}
+}
+
+func (k kobjs) MarshalLog() interface{} {
```

Comment:
What I don't like about this implementation is the reliance on support for MarshalLog. A logger which only supports String will not work well.

Would it be better to implement String? But then what should the result of that be? It cannot be the same `["foo/bar", "x/y"]` that klog currently prints because it has to be a string, not a slice of strings.

Reply:
Implement both and document that MarshalLog is the more performant path.  Encourage loggers to support it.

Reply:
Picking MarshalLog over String does not make sense for all loggers. For example, for klog text format, String is better because the goal is readability.

Reply:
Right now, klog prints `pods=[foo/bar x/y]` because that is what `fmt.Sprintf("%v")` produces. I've implemented `String` so that it returns the same, so the only difference is that there are additional quotation marks: `pods="[foo/bar x/y]"`. This seems reasonable to me.

This means we don't need `MarshalLog` support in klog.

Reply:
> Picking MarshalLog over String does not make sense for all loggers. For example, for klog text format, String is better because the goal is readability.

I'm not sure I agree.  If I implement `MarshalLog()` I am explicitly expressing a preference.  I am not going to accidentally imnplement that for one reason and then find this as a side-effect (which DOES happen with `String()`)

Reply:
But that's exactly what we did with ObjectRef? It implements String and MarshalLog, and we want klog text output to use String and zapr to use MarshalLog.

Implementing *only* MarshalLog is a different case. But as we haven't said that logger implementations must support it, a type that does that will work poorly with those which don't support it - I'd prefer to not have such types.

Reply:
OK, it's a fair point, but I wonder how much of that is "legacy". If we were doing it all from scratch, would we really want to render ObjectRef differently in text vs. structured?

We're kind of trying to read the user's mind here.  MarshalLog is pretty unambiguous, and it seems legit to say (for a specific log impl) "If you set --log-format=text we prefer `String()` methods over `MarshalLog()` methods.  If you set --log-format=struct we prefer `MarshalLog()` over `String()`".

But if you implement `MarshalLog()` which returns a type that also implements `MarshalLog()`, you should do the recursion yourself.  The log impl gives you one chance.

Likewise, if you implement `MarshalLog()` which returns a type that implements `String()`, you should call `String()` yourself.